src/cmd/flux-jobs.py

```diff
@@ -60,6 +60,13 @@ class FluxJobsConfig(UtilConfig):
                 "{priority:<12} {state:<8.8} {dependencies}"
             ),
         },
+        "endreason": {
+            "description": "Show why each job ended",
+            "format": (
+                "{id.f58:>12} ?:{queue:<8.8} {username:<8.8} {name:<10.10+} "
+                "{status_abbrev:>2.2} {t_inactive!D:>19h} {inactive_reason}"
```

Comment:
Hmmm, I think the reason I went with `!D` may have been because we (I don't think) support the `h` hyphen output with `!d`.  So I sort of dislike this output for the running jobs.

```
       JOBID USER     NAME       ST   T_INACTIVE INACTIVE-REASON
    f7qUkw7d achu     sleep       R  Dec31 16:00 
    f7qRnxYx achu     sleep       R  Dec31 16:00 
    f7qRnxYw achu     sleep       R  Dec31 16:00 
    f7qQJyGb achu     sleep       R  Dec31 16:00 
    f7gRwKMZ achu     sleep       F  May10 15:10 User defined signal 1
    f7XHBkU7 achu     sleep       F  May10 15:10 Terminated
    f7GqYkFH achu     sleep      CA  May10 15:10 Canceled: foobar
    f72GVocj achu     sleep      CA  May10 15:10 Canceled
    f6tkdSmM achu     nosuchcom+  F  May10 15:10 command not found
    f6mBo7MH achu     false       F  May10 15:10 Exit 1
    f6dU4rF9 achu     true       CD  May10 15:10 Exit 0
```

Maybe we can tweak/handle `h` in `UtilDatetime`.

But along the way, I noticed this:

```
        if conv == "d":                                                                                                                     
            # convert from float seconds since epoch to a datetime.                                                                         
            # User can than use datetime specific format fields, e.g.                                                                       
            # {t_inactive!d:%H:%M:%S}.                                                                                                      
            try:                                                                                                                            
                value = UtilDatetime.fromtimestamp(value)                                                                                   
            except TypeError:                                                                                                               
                if orig_value == "":                                                                                                        
                    value = UtilDatetime.fromtimestamp(0.0)                                                                                 
                else:                                                                                                                       
                    raise                                                                                                                   
        elif conv == "D":                                                                                                                   
            # As above, but convert to ISO 8601 date time string.                                                                           
            try:                                                                                                                            
                value = datetime.fromtimestamp(value).strftime("%FT%T")                                                                     
            except TypeError:                                                                                                               
                if orig_value == "":                                                                                                        
                    value = ""                                                                                                              
                else:                                                                                                                       
                    raise
```

I don't know why `!d` converts to epoch time with empty string and with `!D` does not.  It is documented that way in the manpage.  That seems inconsistent.  And I think empty string and 0 mean two different things.  For example, `t_remaining` being empty string might mean the job isn't running vs 0 means there's 0 time remaining.  Similarly `t_inactive` being empty string may mean the job isn't done.

I will note that in `JobInfo` we do default:

```
    #  Default values for job properties.
    defaults = {
        "t_depend": 0.0,
        "t_run": 0.0,
        "t_cleanup": 0.0,
        "t_inactive": 0.0,
```

I think it has been this way forever, but maybe shouldn't be.  Wondering if we could default that to `""` and that means output "" vs epoch.

I will note that I'm not sure what fallout there could be with the json output option if we adjusted things.

Will ponder.

Reply:
> I don't know why !d converts to epoch time with empty string and with !D does not. It is documented that way in the manpage. That seems inconsistent. 

One must be a Python datetime object and the other is a string. You cant return an empty string as a datetime object.

Reply:
> Hmmm, I think the reason I went with !D may have been because we (I don't think) support the h hyphen output with !d. So I sort of dislike this output for the running jobs.

Well, this is a bit unsatisfying, but this allows `UtilDatetime` class to use `h`. It requires subverting `h` handling in `format_items` becuase the `UtilDatetime` formatting is done _after_ this function, then adds `h` suffix handling specifically to the `format()` method of the class.

We maybe should also have `datetime.fromtimestamp(0.)` format to an empty string which would help here:

```diff
diff --git a/src/bindings/python/flux/util.py b/src/bindings/python/flux/util.py
index befc950b4..87deb55b6 100644
--- a/src/bindings/python/flux/util.py
+++ b/src/bindings/python/flux/util.py
@@ -312,16 +312,21 @@ class UtilDatetime(datetime):
     def __format__(self, fmt):
         # The string "::" is used to split the strftime() format from
         # any Python format spec:
-        vals = fmt.split("::", 1)
+        timefmt, *spec = fmt.split("::", 1)
 
         # Call strftime() to get the formatted datetime as a string
-        result = self.strftime(vals[0])
+        result = self.strftime(timefmt)
+
+        spec = spec[0] if spec else ""
+        if spec.endswith("h"):
+            if self == datetime.fromtimestamp(0.):
+                result = "-"
+            spec = spec[:-1] + "s"
 
         # If there was a format spec, apply it here:
-        try:
-            return f"{{0:{vals[1]}}}".format(result)
-        except IndexError:
-            return result
+        if spec:
+            return f"{{0:{spec}}}".format(result)
+        return result
 
 
 def fsd(secs):
@@ -421,7 +426,7 @@ class UtilFormatter(Formatter):
             denote_truncation = True
             spec = spec[:-1]
 
-        if spec.endswith("h"):
+        if spec.endswith("h") and not isinstance(value, UtilDatetime):
             localepoch = datetime.fromtimestamp(0.0).strftime("%FT%T")
             basecases = ("", "0s", "0.0", "0:00:00", "1970-01-01T00:00:00", localepoch)
             value = "-" if str(value) in basecases else str(value)
```

Reply:
If this is acceptable I can propose a small PR. It would be more satisfying to handle the `h` in one place, but due to the way the formatting percolates through all the layers here (which frankly is warping my hippocampus).

Reply:
Ahhh your post is very timely, I was looking into how to add `h` and fumbling and just about to message you if you had an idea on how to proceed.

`+        if spec.endswith("h") and not isinstance(value, UtilDatetime):`

This was the part I was struggling to get around, didn't think to just skip it.

PR sounds good.  Thanks.

Reply:
In addition to the changes to the `UtilDatetime` class to support the `h` suffix, what do you think about formatting all "zero" `UtilDatetime` objects as an empty string in the class' `format()` function @chu11?

Seems like this would be consistent with `!D` as you note, and would avoid the requirement to use `h` when you don't want to display `Dec31 ...` for unset timestamps.

Reply:
> Seems like this would be consistent with !D as you note, and would avoid the requirement to use h when you don't want to display Dec31 ... for unset timestamps.

That sounds like a good idea!